azurelinuxagent/common/osutil/default.py

```diff
@@ -1203,25 +1204,42 @@ def device_for_ide_port(self, port_id):
-                    if guid[0] == g0 and guid[1] == "000" + ustr(port_id):
-                        for root, dirs, files in os.walk(path + vmbus):
-                            if root.endswith("/block"):
-                                device = dirs[0]
-                                break
-                            else:
-                                # older distros
-                                for d in dirs:
-                                    if ':' in d and "block" == d.split(':')[0]:
-                                        device = d.split(':')[1]
-                                        break
+        # We have to try device IDs for both Gen1 and Gen2 VMs.
+        for device_id_or_prefix in [
+                '{0}-000{1}'.format(g0, port_id),
+                GEN2_DEVICE_ID]:
+            logger.info('Looking for device ID with name or prefix {0}'.format(device_id_or_prefix))
+            if os.path.exists(path):
+                try:
+                    for vmbus in os.listdir(path):
+                        deviceid = fileutil.read_file(os.path.join(path, vmbus, "device_id"))
+                        guid = deviceid.strip('{}')
+                        if guid.startswith(device_id_or_prefix):
+                            for root, dirs, files in os.walk(path + vmbus):
+                                root_path_parts = root.split('/')
+                                # For Gen1 VMs we only have to check for the block dir in the
+                                # current device. But for Gen2 VMs all of the disks (sda, sdb,
+                                # sr0) are presented in this device on the same SCSI controller.
+                                # Because of that we need to also read the LUN. It will be:
+                                #   0 - OS disk
+                                #   1 - Resource disk
+                                #   2 - CDROM
+                                if root_path_parts[-1] == 'block' and (
+                                        device_id_or_prefix != GEN2_DEVICE_ID or
+                                        root_path_parts[-2].split(':')[-1] == '1'):
+                                    device = dirs[0]
+                                    break
+                                else:
+                                    # older distros
+                                    for d in dirs:
+                                        if ':' in d and "block" == d.split(':')[0]:
+                                            device = d.split(':')[1]
+                                            break
+                            break
+                    if device:
                         break
-            except OSError as oe:
-                logger.warn('Could not obtain device for IDE port {0}: {1}', port_id, ustr(oe))
+                except OSError as oe:
+                    logger.warn('Could not obtain device for IDE port {0}: {1}', port_id, ustr(oe))
```

Comment:
Please refactor this into a helper function to reduce indent levels and simplify understanding of the top-level code. This also isolates gen detection and handling so that, when Gen3 is (inevitably, someday) introduced, it'll be easier and cleaner to handle.

Reply:
This comment caused me to think a little more about my approach. I have refactored it again so that we now only make a single pass on the vmbus dir, so this should be more efficient. It is also greatly simplified, and the indentation and function remains very similar in structure and functionality (with additional gen2 handling, of course) than from before this code change.

Let me know if you think this works, but it should satisfy your recommendation.

Reply:
I still think the inner code should be refactored to a separate method. Even if it only gets called once, from a maintenance perspective, it's better to pull it out. Also, if gen1 and gen2 behaviors diverge more, or if some future gen3 has even more divergent behavior here, you'll want those methods separated into helper functions.

Reply:
Ok, I pulled that code out to a separate method. Thanks!